nixos/modules/services/databases/neo4j.nix

```diff
@@ -36,114 +89,541 @@ let
+        for further details.
+      '';
     };
 
-    extraServerConfig = mkOption {
-      description = "Extra configuration for neo4j server.";
-      default = "";
-      type = types.lines;
+    udc = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the Usage Data Collector which Neo4j uses to collect usage
+          data. Refer to the operations manual section on the
+          <link xlink:href="https://neo4j.com/docs/operations-manual/current/configuration/usage-data-collector/">Usage Data Collector</link>
+          for more information.
+        '';
+      };
     };
+
   };
 
   ###### implementation
 
-  config = mkIf cfg.enable {
-    systemd.services.neo4j = {
-      description = "Neo4j Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "network.target" ];
-      environment = {
-        NEO4J_HOME = "${cfg.package}/share/neo4j";
-        NEO4J_CONF = "${cfg.dataDir}/conf";
-      };
-      serviceConfig = {
-        ExecStart = "${cfg.package}/bin/neo4j console";
-        User = "neo4j";
-        PermissionsStartOnly = true;
-        LimitNOFILE = 40000;
-      };
-      preStart = ''
-        mkdir -m 0700 -p ${cfg.dataDir}/{data/graph.db,conf,logs}
-        ln -fs ${serverConfig} ${cfg.dataDir}/conf/neo4j.conf
-        if [ "$(id -u)" = 0 ]; then chown -R neo4j ${cfg.dataDir}; fi
-      '';
-    };
+  config =
+    let
+      policyNameList = (attrNames cfg.ssl.policies);
+      validPolicyNameList = ["legacy"] ++ policyNameList;
+      validPolicyNameString = concatStringsSep ", " validPolicyNameList;
+    in
+
+    mkIf cfg.enable {
+      assertions = [
+        { assertion = !(elem "legacy" policyNameList);
+          message = "The policy 'legacy' is special to Neo4j, and its name is reserved."; }
+        { assertion = elem cfg.bolt.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.bolt.sslPolicy = ${cfg.bolt.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+        { assertion = elem cfg.https.sslPolicy validPolicyNameList;
+          message = "Invalid policy assigned: `services.neo4j.https.sslPolicy = ${cfg.https.sslPolicy}`, defined policies are: ${validPolicyNameString}"; }
+      ];
+
+      systemd.services.neo4j = {
+        description = "Neo4j Daemon";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        environment = {
+          NEO4J_HOME = "${cfg.package}/share/neo4j";
+          NEO4J_CONF = "${toString cfg.directories.home}/conf";
+        };
+        serviceConfig = {
+          ExecStart = "${cfg.package}/bin/neo4j console";
+          User = "neo4j";
+          PermissionsStartOnly = true;
+          LimitNOFILE = 40000;
+        };
+
+        preStart = ''
+          # Directories Setup
+          mkdir -m 0700 -p ${toString cfg.directories.home}/{certificates,conf,data,import,logs,plugins}
```

Comment:
This needs to be adjusted to the options for the certificate, data, import and plugins paths. Probably just not creating those when they're not the default values.

A neat way of knowing whether the default value is used (without having to do string comparison) is to add `options` to the argument list at the top, then checking `options.services.neo4j.directories.certificates.highestPrio >= 1500`. This value `1500` comes from https://github.com/NixOS/nixpkgs/blob/f682ff93a2778f101d93b68c97278f902523758a/lib/modules.nix#L496 which is used for the `default` attribute in option declarations.

A filter for that over `builtins.attrValues options.services.neo4j.directories` should then get you a list of directories to create. (use `.value` to get the value of an option, `config.some.option == options.some.option.value`)